[WIP] Common feed format

porglezomp opened the pull request:
This is a first attempt at defining a common structure for both Atom and RSS feeds. It's not complete yet, and it doesn't provide the best round trip guarantees right now. Feedback is very much appreciated.

In addition to looking at the docs for both [atom_syndication](https://vtduncan.github.io/rust-atom/atom_syndication/index.html) and [rss](https://frewsxcv.github.io/rust-rss/rss/index.html), the information at [RSS and Atom Compared](http://www.intertwingly.net/wiki/pie/Rss20AndAtom10Compared#head-018c297098e131956bf394c0f7c8b6dd60f5cf78) has been extremely helpful while implementing this.

porglezomp committed:
Add struct definition for common feed format

porglezomp committed:
Add partial support for converting from/to Atom

porglezomp committed:
Add the Entry type

porglezomp committed:
Add conversions between our Entry and Atom's Entry

Use the Entry conversions in the Atom Feed conversions

porglezomp committed:
Fix date parsing to specifically parse RFC 3339

porglezomp committed:
Image field and comment fixes

porglezomp committed:
Keep a clone of the original data in the Feed

porglezomp committed:
Add FromStr for Feed

porglezomp committed:
Split the library into modules

porglezomp committed:
Implement all types needed for the public API

The public API shouldn't expose any types from rss or atom_syndication,
so this adds all the necessary types.

porglezomp commented:
Do we want to expose only the common format as the public API?

tomshen commented:
Thanks for the contribution!

We should only expose the common format as the public API. The purpose of this library is to define a common format so that you can use the same code for working on both Atom and RSS feed, but I just never got around to finishing it. Exposing the Atom/RSS library-specific formats defeats that purpose.

I'll try to take a look at the code sometime this week.

tomshen commented on Cargo.toml:
```diff
@@ -1,13 +1,14 @@
 name = "syndication"
-version = "0.1.1"
-authors = ["Tom Shen <tom@shen.io>"]
+version = "0.1.2"
```
You should rebase your branch, since the library is already on version "0.2.0." This change should bump the version to "0.3.0," since it breaks the original API.

porglezomp commented:
Right now I'm considering doing

``` rust
use atom_syndication as atom;
```

since it gives better symmetry with `rss`, and it's less annoying since I have to disambiguate the types a lot in some of the modules. Is this okay, or should I avoid this?

tomshen commented:
That looks fine to me.

porglezomp committed:
Change Into impls to From impls

porglezomp committed:
Add RSS Entry conversions

Add Guid type since the RSS Guid is more detailed than the Atom one.

porglezomp committed:
Add conversions between our Entry and RSS's Entry

Add the Image and TextInput types to support the change.

porglezomp committed:
Derive Debug, PartialEq, Eq, & Clone when possible

EntryData and FeedData need an explicitly implementation of Debug
because the atom types don't support it, and we probably don't want to
show the contained feeds anyway.

porglezomp committed:
Add some basic tests for Feed

porglezomp commented:
There aren't very good tests for this library, are there. Trying to write some, I guess I can see why. Any ideas for what I should do on the testing front?

porglezomp commented:
Should I put myself as an author?

tomshen commented:
> There aren't very good tests for this library, are there. Trying to write some, I guess I can see why. Any ideas for what I should do on the testing front?

You could use tests from other open source libraries. e.g. https://github.com/kurtmckee/feedparser/tree/develop/tests

> Should I put myself as an author?

Definitely, especially since after this PR merges, most of the codebase will be your work.

Honestly, I haven't looked at Rust stuff in a really long time. So if you want, I can just transfer ownership of the repo/[crate](https://crates.io/crates/syndication) to you.

porglezomp committed:
Merge branch 'master' into HEAD

porglezomp committed:
Update version number and add myself as an author

porglezomp commented:
I just finished up classes for the year, and I accidentally broke my main RSS reader, so I figure now's a good time to come back to this. I'm going to try to use this to build a small feed reader in order to figure out if the API has any glaring holes.

nc7s commented:
Hi, is there still motivation for this? I'm currently writing an application that requires unified RSS & Atom parsing, this would be a great help. Willing to pick it up if there's no time for this on your side.

porglezomp commented:
My job means I'm currently not able to easily make open source contributions, I would be happy with someone else picking up the work.
